CMakeLists.txt

```diff
@@ -459,6 +459,12 @@ try_append_cxx_flags("-fstack-reuse=none" TARGET core_interface)
 if(ENABLE_HARDENING)
   add_library(hardening_interface INTERFACE)
   target_link_libraries(core_interface INTERFACE hardening_interface)
+
+  target_compile_options(hardening_interface INTERFACE
+    -D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_DEBUG
+    -D_GLIBCXX_ASSERTIONS
+  )
+
```

Comment:
How is this related to the following?

https://github.com/bitcoin/bitcoin/blob/e8cc790fe2a223cfaa6dedfba88a6dc4f336f1d9/depends/hosts/linux.mk#L16-L20

Reply:
These are flags set for `DEBUG=1` when building depends. Iirc `-D_GLIBCXX_DEBUG` includes `-D_GLIBCXX_ASSERTIONS` as well as other slower checks.

Reply:
> How is this related to the following?

I understand that the flags in `bitcoin/depends/hosts/linux.mk` are (obviously) only used when building depends and only on Linux. In comparison, if added to `CMakeLists.txt` under `ENABLE_HARDENING` then the flags will be used for depends and for non-depends builds on all platforms that use libc++ or libstdc++, not only Linux. Thus I dropped the flags from `bitcoin/depends/hosts/linux.mk` and added (a more extensive) hardening to `CMakeLists.txt`.

Yes, `-D_GLIBCXX_DEBUG` includes `-D_GLIBCXX_ASSERTIONS`.

> How do they interact?

I guess it is harmless to specify both `_GLIBCXX_DEBUG*` and `_LIBCPP_HARDENING_MODE` because libstdc++ will honor the first and libc++ the second. 

> I don't think it is useful to possibly silently downgrade DEBUG into FAST.

Indeed that would be bad. If `_LIBCPP_HARDENING_MODE` is provided more than once on the command line, then a warning will be raised: "macro is redefined". E.g. `-D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_DEBUG -D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_FAST` will produce a warning.

> Also, it can be considered to enable the ABI checks

Done.

Reply:
> I don't think it is useful to possibly silently downgrade DEBUG into FAST.

> If _LIBCPP_HARDENING_MODE is provided more than once on the command line, then a warning will be raised

Just printing a warning will more than likely still result in silent downgrading.

Reply:
`-Werror` would flip that warning into an error. But anyway, in the latest incarnation of this PR `_LIBCPP_HARDENING_MODE` is set from one place. So there is no way that one part of the build system setting it to one value and another part of the build system setting it to another value, overriding the first one.

Reply:
The libc hardening is now removed from `depends/hosts/linux.mk` in favor of more universal hardening from `CMakeLists.txt` (which is enabled when building via depends).

Reply:
How would a vendor enable `_LIBCPP_HARDENING_MODE`? Some way other than `cmake -DENABLE_HARDENING=ON`? If they supply flags somehow externally, then the same argument can be made for any flag - our build system may override a flag set externally (e.g. `-O1` vs `-O2` or whatever).

Thinking about this, the proper way to supply flags externally is to use `cmake -DAPPEND_CPPFLAGS="-DFOO=1"`.

I think this is safe:

* The proper way to enable `_LIBCPP_HARDENING_MODE` and other hardening options depending on compiler and environment is via `cmake -DENABLE_HARDENING=ON`.
  * If somebody uses `cmake -DENABLE_HARDENING=ON` _and_ supplies `_LIBCPP_HARDENING_MODE` externally, then they are shooting themselves in the foot and will get compilation warning or error depending on `-Werror` about redefined macro. The warnings will be all over the place, for every invocation of the compiler for every file.
* It would be possible to fiddle manually with this by using `cmake -DENABLE_HARDENING=OFF -DAPPEND_CPPFLAGS="-D_LIBCPP_HARDENING_MODE=XYZ"`

Reply:
> Thinking about this, the proper way to supply flags externally is to use cmake -DAPPEND_CPPFLAGS="-DFOO=1".

No? The append_* flags are a hack we had to add to bypass various other CMake behaviours and they are not intended for general use (hence "special builds").

Reply:
> I meant a libc++ vendor

Oh, sorry for misunderstanding.

So, according to https://libcxx.llvm.org/Hardening.html#notes-for-users the libc++ vendor can set a default value and users that wish to use a different value can do that by passing `-D_LIBCPP_HARDENING_MODE=...` to the compiler which is exactly what Bitcoin Core will do. Sounds good?

Reply:
Using an example would help. Are you worried about the following?

A libc++ vendor, lets say "Ubuntu 123.45 LTS", compiles the libc++ they ship with `LIBCXX_HARDENING_MODE=debug` (see https://libcxx.llvm.org/Hardening.html#notes-for-vendors). Then a user of "Ubuntu 123.45 LTS" compiles Bitcoin Core with `cmake -DCMAKE_BUILD_TYPE=Release`. This results in Bitcoin Core being compiled with `fast` hardening mode.

The above is how it should be - the OS default is supposed to be overridable and in Bitcoin Core `-DCMAKE_BUILD_TYPE=Release` is intended to use the `fast` mode.

There is also this note in the docs:

> Since the static and shared library components of libc++ are built by the vendor, setting this macro will have no impact on the hardening mode for the pre-built components. Most libc++ code is header-based, so a user-provided value for _LIBCPP_HARDENING_MODE will be mostly respected.

I think it is no-brainer that an application (e.g. Bitcoin Core) should pass `-D_LIBCPP_HARDENING_MODE=...` to the compiler and that is the correct and the best thing an app can do and this PR does that.

Reply:
Extended the PR description, see the <sup>1</sup> specifically. I wouldn't call it a "downgrade" - it is less checks, but an "upgrade" from performance point of view. All these settings provide varying tradeoffs between performance and amount of checks. Downgrade for one is upgrade for the other.

This is analogous to the `-O` flag. Assuming the default is `-O1` (if no `-O...` is provided). If the user explicitly provides `-O2` that will use `-O2` instead of the default `-O1`. I wouldn't call this (silent) "downgrade" or "upgrade", but rather "intentional override of the default".